### PR TITLE
Budget App

### PR DESCRIPTION
**Pull request by Izebiafe**
# Project Description - Budget App

## Overview
The Budget app project focuses on building a mobile web application for managing personal budgets. The application allows users to track transactions associated with different categories, providing insights into their spending habits.

### User Authentication
- Users can register and log in to access personalized data securely.

### Transactions Management
- Users can add new transactions associated with specific categories.
- View a list of transactions and the total amount spent for each category.

### Design Guidelines
- Adhere to design guidelines, including colors, typography, and layout.
- Follow the provided design idea by Gregoire Vella on Behance with a Creative Commons license.

#### Splash Screen
- Simple page displaying the app's name and links to sign-up and log-in pages.

#### Sign-Up and Log-In Pages
- User registration with full name, email, and password (mandatory fields).
- User login using email and password.
- Restrict access to pages requiring authentication when the user is not logged in.

#### Home Page (Categories Page)
- Categories page displaying category names, icons, and total transaction amounts.
- Clicking on a category navigates to the transactions page for that category.
- "Add a new category" button for creating a new category.

#### Transactions Page
- List of transactions for a selected category, ordered by the most recent.
- Display the total amount for the category at the top.
- "Add a new transaction" button for creating a new transaction.

#### "Add a New Category" Page
- Form to create a new category with mandatory fields (name and icon).
- "Save" button to create the new category and navigate to the home page.

#### "Add a New Transaction" Page
- Form to create a new transaction with mandatory fields (name, amount, and at least one category).
- "Save" button to create the new transaction and navigate to the transactions page for that category.

### Testing Requirements
- Develop unit and integration tests for critical components of the Rails application.

### Technical Stack
- Database: Postgres.
- Authentication: Devise.
- Validate all user input to enhance app security.
- Use a view template engine of choice (.erb, .slim, .haml).

### Deployment
- Deploy the application online for accessibility.

### Database Schema
- Reflect the provided ERD diagram structure.

## Project Documentation
- Record a 5-minute video presenting the project's features.
- Include a description of the project, a demonstration of features, and highlight interesting pieces of code.

## Requirements and Best Practices
- Ensure no linter errors.
- Follow correct gitflow.
- Document work professionally.
- Adhere to best practices for Ruby.
- Use Devise for authentication.

## Conclusion
This Ruby on Rails capstone project combines user-friendly interactions, aesthetic design, and robust authentication to create a comprehensive budget management web application. The adherence to best practices and thorough testing ensures a reliable and secure user experience.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,169 +1,169 @@
...
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+<!-- LIVE DEMO -->
+
+## 🚀 Live Demo & Presentation <a name="live-demo"></a>
+
+- [Live Demo Link](https://budget-app-dr4e.onrender.com)
+<!--- [Video presentation]https://drive.google.com/file/d/12L3Cbr-AQqXLpPbJ8xi-hcw6PMuangFW/view?usp=sharing -->
+
```
Comment (Ibizugbe) [inline README.md]: Nice job so far😉. Kindly ensure the live version of your project is properly deployed without any bugs, as it currently renders a broken app😉. Please do well to fix this, as this is a requirement that our project is properly deployed and everything works fine👍🏼

image


Also, as a requirement, you are meant to record a video presentation showing your project and add the link here, please do well to add that 😉.

image

**Review by Ibizugbe — CHANGES_REQUESTED** [review]
Hi @Izebiafe 👋🏼👋🏼👋🏼, 

request change


Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights✅✅✅✅

- [x] Linters are passing👍🏼.
- [x] All tests are passing👍🏼.
- [x] Routes are protected👍🏼.
- [x] Inputs are validated👍🏼.
- [x] Nice job adding the edit and delete options for each transaction👍🏼.
- [x] Devise is used for authentication👍🏼.


## Required Changes ♻️

- [ ] Nice job so far👍🏼. Please let's try to stick to the design for the slash screen😉. Kindly add the name of your app at the appropriate position (center) instead of the categories link as you implemented😉. Also, try to ensure the buttons and the entire page match the design in colors and spacing😉.

yours
image

design


image


_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Ibizugbe — COMMENTED** [review]
Hi @Izebiafe 👋🏼👋🏼👋🏼, 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

You haven't made any changes since the last review🥲 Please do well to fix the issue as shown in the previous review👍🏼. 

### **Please do well to also add a video presentation to your project as this is also a requirement and I forgot to mention this in the previous review😉😉**

image


Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

File: README.md [hunk 2]
```diff
@@ -1,169 +1,169 @@
...
+- **Categorize purchases to see exact amounts spent on groceries, eating out, shopping, bills etc.**
+- **Set custom budgets for each category and track over/under spending**
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+<!-- LIVE DEMO -->
+
+## 🚀 Live Demo & Presentation <a name="live-demo"></a>
+
+- [Live Demo Link](https://budget-app-dr4e.onrender.com)
```
Comment (mckent05) [inline README.md]: - You have done well so far, but as the last reviewer said, please ensure that your application is deployed properly, currently, I am still having the same errors when I try to use your application online. Please fix this 👍 👍 
wwefqwefqwef

File: README.md [hunk 3]
```diff
@@ -1,169 +1,169 @@
...
+If you like this project consider adding a ⭐️
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+<!-- ACKNOWLEDGEMENTS -->
+
+## 🙏 Acknowledgments <a name="acknowledgements"></a>
+
+- I would like to thank Microverse.
+- Original design idea by: Gregoire Vella on Behance.
```
Comment (mckent05) [inline README.md]: **OPTIONAL** - You did well acknowledging the authors of the original design, but it'd be better if you added a link leading to the design portfolio so future hiring managers can see the design and tried to replicate, please fix this 👍 👍

**Review by mckent05 — CHANGES_REQUESTED** [review]
Hi @Izebiafe, 

# STATUS: CHANGES REQUIRED ♻️ ♻️ 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
almost there

## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your PR is professional 👍 👍 
- You did well working on your capstone project 👍 👍 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: README.md [hunk 4]
```diff
@@ -1,169 +1,169 @@
...
+- **Categorize purchases to see exact amounts spent on groceries, eating out, shopping, bills etc.**
+- **Set custom budgets for each category and track over/under spending**
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+<!-- LIVE DEMO -->
+
+## 🚀 Live Demo & Presentation <a name="live-demo"></a>
+
+- [Live Demo Link](https://budget-app-stqp.onrender.com)
```
Comment (danmainah) [inline README.md]: - [ ] @Izebiafe Great work fixing the error preventing the deployed app from opening, I am able only to access the splash screen, when I click the Sign in or Sign up button I get the error below, kindly check the application logs to identify the issue causing this.
image

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @Izebiafe , 

Great work on making the changes requested by a previous reviewer 👏  👏 👏
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
✔️ Implemented the change requested to add a link to the design portfolio.
✔️ Implemented the change to fix the 502 error when trying to open the live demo application.


 # Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: README.md (hunk 4 above)
Comment (Izebiafe) [reply README.md]: @danmainah it works well at my end

Comment (shahadat3669) [inline README.md]: Hey @Izebiafe, I wanted to thank you for fixing the error that was preventing the deployed app from opening. However, I noticed that when I try to access the app from the provided link, I get an error saying 'Not Found'. I have attached a video to show you what I mean. Can you please verify and provide a working link in the readme file. Wish you good luck!


https://github.com/Izebiafe/Budget-app/assets/55840999/0245ca41-3688-4047-bd85-965e933ad336

**Review by shahadat3669 — CHANGES_REQUESTED** [review]
Hi @Izebiafe, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.




## Don't Worry 🙏

It's okay to have bugs in your code. 🆗 Please see this code review as an opportunity to learn practical coding skills and discover new tips that can improve your code quality 💯 and help you **get even better** 💼.

**I will guide you on how to address these areas down below 👇.**

## Highlights
✅ Worked on the previous reviewer comments

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

## Have any doubt ❓

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

<table  width="100%" align="center">
  <tr>
    <td colspan="4" align="center">
      <strong>Contact me: Shahadat Hossain</strong>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://github.com/shahadat3669">
        GitHub
      </a>
    </td>
    <td align="center">
      <a href="https://www.linkedin.com/in/shahadat-cseng/">
        LinkedIn
      </a>
    </td>
    <td align="center">
      <a href="https://app.slack.com/client/T47CT8XPG/D05GT4E4WAY/rimeto_profile/U04426SD71V">
        Slack
      </a>
    </td>
    <td align="center">
      <a href="mailto:shahadat3669@gmail.com?subject=Microverse%20Code%20Review">
        Email
      </a>
    </td>
  </tr>
</table>

## ⚠️ WARNING ⚠️

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by LionRouge1 — APPROVED** [review]
Hi @Izebiafe, 

# STATUS: APPROVED! 🥇 💯 
Great work on making the changes 👏 
You've done well implementing the requested changes

done

Your project is complete! There is nothing else to say other than... it's time to merge it :shipit:

## Highlights
- Great job, your project is error free :+1: 
- Nice work you meet all the requirements :1st_place_medal: 
- All linters passed :heavy_check_mark: 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Good luck! 🎉 


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**